### PR TITLE
Validate GridSample input spatial dimensions are non-empty

### PR DESCRIPTION
**Pull request by titaiwangms**
### Description

`GridSample` sizes its output from the grid, so an input with a zero-size spatial dimension skips the empty-output early return and reaches interpolation with an invalid spatial extent. This change validates that every input spatial dimension is non-empty in both the CPU and CUDA kernels, returning a clear error for such malformed inputs.

### Changes

- CPU `GridSample::Compute`: reject any zero-size input spatial dimension with a descriptive error.
- CUDA `GridSample::ComputeInternal`: same validation, layout-aware for NCHW and NHWC, covering both 4-D (H, W) and 5-D (D, H, W) inputs. Messages are consistent across both EPs.
- Add an expect-failure test (zero-size spatial dim, `padding_mode=border`) scoped to the CPU and CUDA execution providers, which are the kernels that carry the guard.

### Motivation

Improves input validation and error diagnostics for malformed `GridSample` inputs. Covers both the CPU and CUDA execution providers; no behavior change for valid inputs.

Co-authored-by: Copilot <223556219+Copilot@users.noreply.github.com>

### CODE REVIEW

**Comment by yuslepukhin** [comment]
Issues found
Inconsistent error handling style — CPU kernel (Minor)

In grid_sample.cc:374-382, the existing checks use ORT_ENFORCE (throws on failure), but the new check uses ORT_RETURN_IF_NOT (returns error Status). This is actually the correct choice for the new code — ORT_RETURN_IF_NOT is preferable for input validation since it returns a clean Status rather than throwing. However, it means that this kernel mixes error-handling styles: the preceding rank/grid-consistency checks all ORT_ENFORCE, which will abort if exceptions are disabled. The new code is fine as-is; the old code is the problem, not this PR.

2. WebGPU EP is also vulnerable (Not addressed) The WebGPU GridSample::ComputeInternal at grid_sample.cc:197-240:

Derives output from grid shape (H_out, W_out)
Has an early return only when output_size == 0

Passes H_in/W_in as shader uniforms — if they are 0, the gs_pixel function's bounds checks (r >= i32(uniforms.H_in)) will treat every coordinate as OOB, which for padding_mode=zeros returns 0 (safe), for padding_mode=border clamps to clamp(r, 0, -1) → undefined behavior in WGSL (or at least incorrect), and for padding_mode=reflection divides by 0 in range = v_max - v_min (e.g., f32(0) - 1.0 = -1.0, then dv / range with negative range).

The PR description acknowledges this with "WebGPU/CoreML use separate kernels without that guard, so the expect-failure test below must exclude them." This is acceptable as a scoped fix but the WebGPU EP should get the same guard in a follow-up.

## Missing tests ##

Zero W instead of zero H — e.g., {1,1,5,0} to verify the loop catches non-first spatial dimensions
All spatial dims zero — {1,1,0,0} to confirm no double-error or crash
5-D with non-first spatial dim zero — e.g., {1,1,2,0,3} or {1,1,2,3,0}
NHWC layout for CUDA — The test runs on DefaultCudaNHWCExecutionProvider() but uses NCHW-shaped input {1,1,0,5}. An explicit NHWC-shaped input (e.g., {1,0,5,1} for the NHWC interpretation) would exercise the IsNHWC ? 1 : 2 / IsNHWC ? size()-1 : size() code path more meaningfully.

padding_mode=zeros — The tests only use border. Adding a zeros variant ensures the guard fires before the interpolation regardless of padding mode.
padding_mode=reflection — Same reason; this is the most dangerous mode (division by zero in reflect logic).
Valid input sanity test — A test with non-zero spatial dims that still passes, confirming no false rejection.

**Review by justinchuby — DISMISSED** [review]

**Review by justinchuby — APPROVED** [review]

**Review by justinchuby — APPROVED** [review]

File: onnxruntime/core/providers/cpu/tensor/grid_sample.cc [hunk 1]
```diff
@@ -379,6 +379,14 @@ Status GridSample<T>::Compute(OpKernelContext* context) const {
 
   ORT_ENFORCE(input_dims.NumDimensions() == 4 || input_dims.NumDimensions() == 5, "Only 4-D or 5-D tensor is supported");
 
+  // Spatial dimensions must be non-empty for sampling: the output is sized by the grid, so a zero-size
+  // input spatial dimension would otherwise lead to invalid index computations during interpolation.
+  for (size_t i = 2; i < input_dims.NumDimensions(); ++i) {
```
Comment (tianleiwu) [inline onnxruntime/core/providers/cpu/tensor/grid_sample.cc:384]: Optional: this guard runs before the empty-output early return further down (`if (Y.Shape().Size() == 0) return Status::OK();`). When the grid yields a zero-size output, a zero-size input spatial dim previously returned OK and now errors. Since no sampling occurs for an empty output, consider moving the check after the early return so genuinely empty outputs still short-circuit, keeping consistency with the existing zero-size-output handling.

File: onnxruntime/core/providers/cuda/tensor/grid_sample.cc [hunk 2]
```diff
@@ -124,6 +124,17 @@ Status GridSample<T, IsNHWC>::ComputeInternal(OpKernelContext* context) const {
     return Status(common::ONNXRUNTIME, common::NOT_IMPLEMENTED, "Cubic mode is only supported in 4-D cases.");
   }
 
+  // Spatial dimensions must be non-empty for sampling: the output is sized by the grid, so a zero-size
+  // input spatial dimension would otherwise lead to invalid index computations during interpolation.
+  // Spatial dims are [1, rank-1) for NHWC (N,...,C) and [2, rank) for NCHW (N,C,...).
+  const size_t spatial_begin = IsNHWC ? 1 : 2;
```
Comment (tianleiwu) [inline onnxruntime/core/providers/cuda/tensor/grid_sample.cc:130]: Same observation as the CPU kernel: this guard runs before the empty-output early return (`if (Y->Shape().Size() == 0) return Status::OK();`). A zero-size input spatial dim combined with an empty grid/output previously returned OK and now errors. Consider placing the check after the early return so empty outputs still short-circuit, since no interpolation happens in that case.

File: onnxruntime/test/providers/cpu/tensor/grid_sample_test_custom.cc [hunk 3]
```diff
@@ -602,10 +613,245 @@ TYPED_TEST(GridSampleCustomTest, test_grid_sample_20_4D_nearest_reflection_mixed
...
+  test.AddInput<TypeParam>("Grid", Grid_shape, Grid_data);
+  test.AddOutput<TypeParam>("Y", Y_shape, {TypeParam(0.0f), TypeParam(0.0f)});
+  // Restrict the expect-failure run to CPU + CUDA only, matching the 4-D case above.
+  RunTestsExpectFailure(test, "Input spatial dimensions must be non-empty for sampling",
+                        GetCpuAndCudaExecutionProviders());
+}
+
+TYPED_TEST(GridSampleCustomTest, test_grid_sample_20_4D_nearest_border_empty_last_spatial_rejected) {
+  // Zero-size trailing spatial dimension (W = 0), confirming the guard loop catches a non-first
+  // spatial dimension, not just the leading one. This runs on the CUDA NHWC provider as well: for a
```
Comment (tianleiwu) [inline onnxruntime/test/providers/cpu/tensor/grid_sample_test_custom.cc:675]: The comment states this test also runs on the CUDA NHWC provider and exercises the layout-aware spatial range. `GetCpuAndCudaExecutionProviders()` uses the default CUDA EP, which runs the NCHW GridSample variant (`IsNHWC=false`). The NHWC kernel is registered only under `kMSInternalNHWCDomain` (behind `ENABLE_CUDA_NHWC_OPS`) and is selected only when the provider's `prefer_nhwc` option triggers a layout transform. So this case validates the NCHW guard catching a trailing zero dimension (still useful), but it does not actually exercise the NHWC path — consider trimming the comment to avoid overstating coverage.

**Review by tianleiwu — COMMENTED** [review]
Overall this is a clean, well-tested input-validation hardening: the non-empty spatial-dimension guard is applied consistently across the CPU, CUDA, and WebGPU kernels with a shared, descriptive error message, and the expect-failure tests cover 4-D/5-D, leading/trailing/middle zero dims, and all padding modes, plus a positive sanity case. A few non-blocking notes below.

1. Placement relative to the empty-output early return (inline on CPU/CUDA): the guard runs before the existing `if (Y.Shape().Size() == 0) return Status::OK();` short-circuit, so a zero-size input spatial dim combined with a zero-size grid/output — which previously returned OK — now errors. Since no sampling happens when the output is empty, consider placing the guard after that early return to preserve the existing zero-size-output behavior. Non-blocking.

2. PR description scope: the "Changes"/"Motivation" sections mention only the CPU and CUDA kernels, but this PR also adds the same guard to `onnxruntime/core/providers/webgpu/tensor/grid_sample.cc` and a WebGPU expect-failure test. Worth updating the description to reflect the WebGPU change so it matches the diff.

(The earlier automated suggestion about returning `INVALID_ARGUMENT` explicitly vs. `ORT_RETURN_IF_NOT` is still a reasonable consistency point but is left to author discretion.)
